docs/machine-learning/basic-concepts-model-training-in-mldotnet.MD

```diff
@@ -0,0 +1,169 @@
+
+It is also common that the input and output column names are the same. In this case, the old column is 'replaced' with the new one. For example, a `new HashTransformer("foo")` would take the values from column "foo", hash them and 'put them back' into "foo". 
+
+Any transformer will, of course, produce a new data view when `Transform` is called: remember, data views are immutable.
+
+Another important consideration is that, because data is lazily evaluated, *transformers are lazy too*. Essentially, after you call
+
+```csharp
+var newData = transformer.Transform(oldData)
+```
+
+no actual computation will happen: only after you get a cursor from `newData` and start consuming the value will `newData` invoke the `transformer`'s transformation logic (and even that only if `transformer` in question is actually needed to produce the requested columns).
+
+### Transformer chains
+
+A useful property of a transformer is that *you can phrase a sequential application of transformers as yet another transformer*:
+
+```csharp
+var fullTransformer = transformer1.Append(transformer2).Append(transformer3);
+```
+
+We utilize this property a lot in ML.NET: typically, the trained ML.NET model is a 'chain of transformers', which is, for all intents and purposes, a *transformer*. 
+
+## Data reader
+
+The data reader is ML.NET component to 'create' data: it takes an instance of `T` and returns data out of it.
+
+Here's the exact interface of `IDataReader<T>`:
+
+```c#
+public interface IDataReader<in TSource>
+{
+    IDataView Read(TSource input);
+    ISchema GetOutputSchema();
+}
+```
+
+As you can see, the reader is capable of reading data (potentially multiple times, and from different 'inputs'), but the resulting data will always have the same schema, denoted by `GetOutputSchema`.
+
+An interesting property to note is that you can create a new data reader by 'attaching' a transformer to an existing data reader. This way you can have 'reader' with transformation behavior baked in:
+
+```c#
+var newReader = reader.Append(transformer1).Append(transformer2)
+```
+
+Another similarity to transformers is that, since data is lazily evaluated, *readers are lazy*: no (or minimal) actual 'reading' happens when you call `dataReader.Read()`: only when a cursor is requested on the resulting data does the reader begin to work.
+
+## Estimator
+
+The *estimator* is an object that learns from data. The result of the learning is a *transformer*.
+Here is the interface of `IEstimator<T>`:
+
+```c#
+public interface IEstimator<out TTransformer>
+    where TTransformer : ITransformer
+{
+    TTransformer Fit(IDataView input);
+    SchemaShape GetOutputSchema(SchemaShape inputSchema);
+}
+```
+
+You can easily imagine how *a sequence of estimators can be phrased as an estimator* of its own. In ML.NET, we rely on this property to create 'learning pipelines' that chain together different estimators:
+
+```c#
```

Comment:
This is the old version -- we'll find an example.

Reply:
Replace code with this:
MLContext mlContext = new MLContext(seed: 0);

var pipeline = new CopyColumnsEstimator(mlContext, "FareAmount", "Label")
             .Append(new OneHotEncodingEstimator(mlContext, "VendorId"))
             .Append(new OneHotEncodingEstimator(mlContext, "RateCode"))
             .Append(new OneHotEncodingEstimator(mlContext, "PaymentType"))
             .Append(new ColumnConcatenatingEstimator(mlContext, "Features", "VendorId", "RateCode", "PassengerCount", "TripTime", "TripDistance", "PaymentType"))
             .Append(mlContext.Regression.Trainers.FastTree("Label", "Features"));

var model = pipeline.Fit(dataView);

OK, @CESARDELATORRE @GalOshri?

Reply:
We should use `.Append(mlContextml.Transforms.Categorical.OneHotEncoding()` instead of `new OneHotEncodingEstimator()`.

Reply:
This is the pattern for all components. They should be available through mlContext.